internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -498,6 +517,65 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, rollback bool, s
 	return cb, nil
 }
 
+func (u *Upgrader) addNewInstallToRegistry(version string, unpackRes UnpackResult, detectedFlavor string) {
+	_, err := u.installDescriptorSource.AddInstallDesc(
+		v1.AgentInstallDesc{Version: version, VersionedHome: unpackRes.VersionedHome, Hash: unpackRes.Hash, Flavor: detectedFlavor, Active: false},
+	)
+	if err != nil {
+		u.log.Warnf("error encountered when adding install description of new agent version: %s", err.Error())
+	}
+}
+
+func (u *Upgrader) activateInstallInRegistry(newVersionedHome, currentVersionedHome string, rollbackWindow time.Duration, rotationTimestamp time.Time) *v1.InstallDescriptor {
+	modifiedInstallDescriptor, err := u.installDescriptorSource.ModifyInstallDesc(
+		func(desc *v1.AgentInstallDesc) error {
+			if desc.VersionedHome == newVersionedHome {
+				desc.Active = true
+				return nil
+			} else {
+				desc.Active = false
+			}
+
+			// set the TTL only for the current install
+			if desc.VersionedHome == currentVersionedHome {
+				desc.TTL = getCurrentInstallTTL(rollbackWindow, rotationTimestamp)
+			}
+
+			return nil
+		},
+	)
+	if err != nil {
+		u.log.Warnf("error encountered when setting new install description as active:  %s", err.Error())
+	}
+	return modifiedInstallDescriptor
+}
+
+func getAvailableRollbacks(rollbackWindow time.Duration, now time.Time, newVersionedHome string, descriptor *v1.InstallDescriptor) []v1.AgentInstallDesc {
+	if rollbackWindow == 0 {
+		// if there's no rollback window it means that no rollback should survive the watcher cleanup at the end of the grace period.
+		return nil
+	}
+
+	res := make([]v1.AgentInstallDesc, 0, len(descriptor.AgentInstalls))
+	for _, installDesc := range descriptor.AgentInstalls {
+		if installDesc.VersionedHome != newVersionedHome && (installDesc.TTL == nil || now.Before(*installDesc.TTL)) {
```

Comment:
- https://github.com/elastic/ingest-dev/issues/6061

I think there is some danger here, if the system went to sleep long enough (consider people turn off their laptops on vacation) the previous install could be detected as expired before launching the watcher for an upgrade. We also want to make sure people can't brick their agents during upgrade if they set a rollback TTL of a very small value accidentally and remove the ability to do an  automated rollback.

Are we able to keep the previous versions around and be able to roll back to it  regardless of what the rollback configuration or TTL might be if there is an active upgrade in progress or something along those lines?